src/reset-password/service.ts

```diff
@@ -25,3 +28,28 @@ export async function sendResetPasswordEmail(user: User) {
 
     nodemailer.getTestMessageUrl(info);
 }
+
+export async function createToken(user: User): Promise<string> {
+    const token = crypto.randomBytes(32).toString('base64url').replace('=', '');
+    const query = 'INSERT INTO reset_password_token SET user_id = ?, token = ?, created_at = UNIX_TIMESTAMP()';
+    await database.query(query, [
+        user.id,
+        await bcrypt.hash(token, 12),
+    ]);
+    return token;
+}
+
+export async function validateToken(user: User, token: string): Promise<boolean> {
+    const query = 'SELECT token FROM reset_password_token WHERE user_id = ?';
+    const result = await database.query(query, [user.id]);
+
+    const hashes: string[] = result[0].map( (row: { token: string }) => row.token );
+
+    for (const hash of hashes) {
+        if (await bcrypt.compare(token, hash)) {
+            return true;
+        }
+    }
```

Comment:
after validating a token, is there any mechanism by which to mark it as validated? i.e., that the account recovery is completed?

Reply:
this should be added

Reply:
a suggestion: change the `created_at = UNIX_TIMESTAMP()` field to `expires_at = UNIX_TIMESTAMP() + TTL` and then, after validation, set `expires_at = UNIX_TIMESTAMP()`

Reply:
Agree with adding a new field, but I think we can just delete the record after it was used?

Reply:
is there another mechanism to audit one's account and see the last time a password was reset or a reset was requested?

Reply:
Yes, we have audit logs and we should add these as new events! good idea. Probably need 2 types of events:

* password reset requested
* password reset completed